runtime/vm/mz64/UpcallThunkGen.cpp

```diff
@@ -59,16 +56,188 @@ extern "C" {
+                }
+                currArgIndex += 1;
+                fprIndex += 2;
+                if (i < 3) {
+                    gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+                }
+                break;
+            case J9_FFI_UPCALL_SIG_TYPE_DOUBLE:
+                if (fprIndex <= maxFPRIndex) {
+                    thunkMem += STDY(thunkMem, fprIndex, 0, R4, argAreaOffsetFromSP + (currArgIndex * 8));
+                }
+                currArgIndex += 1;
+                fprIndex += 2;
+                if (i < 3) {
+                    gprIndex++; // If one of the first three doublewords of the argument area is float/double, then the corresponding GPR is undefined. So we must ignore any value in there.
+                }
+                break;
+            default:
+                Assert_VM_unreachable();
+        }
+
+    }
+
+    // Now jump to the common dispatcher helper routine.
+    thunkMem += IIHF(thunkMem, R1, ((reinterpret_cast<uint64_t>(metaData)) >> 32));
+    thunkMem += IILF(thunkMem, R1, reinterpret_cast<uint64_t>(metaData));
+    thunkMem += LAY(thunkMem, R2, 0, R4, argAreaOffsetFromSP); // argument pointer
+    uint64_t upcallAddress = *(reinterpret_cast<uint64_t *>(reinterpret_cast<uint64_t>(metaData->upCallCommonDispatcher) + 8));
+    // Materialize the destination address in R6.
+    thunkMem += IIHF(thunkMem, R6, upcallAddress >> 32);
+    thunkMem += IILF(thunkMem, R6, upcallAddress);
+
+    // Now generate the branch to the helper
+    thunkMem += BASR(thunkMem, R7, R6);
+
+    // Mandatory NOP instruction
+    // With XPLINK, the function entry point address is not always passed to the called function. To allow Language Environment and other tools to find the entry point of the currently executing routine, every
+    // call site, which is located by the "return address" field of the current stack frame, contains information necessary to locate the entry points of both the calling and called functions and, if required,
+    // information about floating-point parameters passed. This is done by encoding information in a no-op instruction at the return point.
+    // This is encoded using a 2–byte NOPR instruction, like the BCR below.
+    thunkMem += BCR(thunkMem, 0, 0);
+    // Upon returning copy back the structure, then restore the stack pointer.
+    thunkMem += LMG(thunkMem, R6, R15, R4, offsetToRegSaveArea + thunkFrameSize); // 1808+256
+    thunkMem += LAY(thunkMem, R4, 0, R4, thunkFrameSize);
+    // Then branch back
+    thunkMem += BC(thunkMem, 0xF, 0, R7, 2);
+
+    // Finish up before returning
+    vmFuncs->doneUpcallThunkGeneration(metaData, (void *)(metaData->thunkAddress));
+    metaData->functionPtr[0] = (UDATA)(*(reinterpret_cast<uint64_t *>(metaData->upCallCommonDispatcher)));
+    metaData->functionPtr[1] = (UDATA )(metaData->thunkAddress);
```

Comment:
Should we use `reinterpret_cast` here for `UDATA` ? Same at line 230.

Reply:
Fixed in 40a47a51532.